### PR TITLE
Submission cost in list submissions API

### PR DESCRIPTION
**Pull request by ghost**
This implements [#15](https://app.zenhub.com/workspace/o/databiosphere/firecloud-app/issues/15)
Also added a test to confirm that we bypass BigQuery when an empty list of workflows is given. 

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala [hunk 1]
```diff
@@ -130,9 +130,14 @@ trait SubmissionComponent {
...
           val config = methodConfigurationQuery.unmarshalMethodConfig(methodConfigRec, Map.empty, Map.empty, Map.empty)
-          val subStatuses = states.getOrElse(submissionRec.id, Seq.empty).map(x => x.workflowStatus -> x.count).toMap
-
-          new SubmissionListResponse(unmarshalSubmission(submissionRec, config, entityRec.map(_.toReference), Seq.empty), user, subStatuses)
+          val wfStates = states.getOrElse(submissionRec.id, Seq.empty).map(x => x.workflowStatus -> x.count).groupBy(_._1) map {
+            case (status, statusToCount) => status -> statusToCount.map(_._2)
+          }
+          val statusCounts = wfStates map {
+            case (status, count) => status -> count.sum
+          }
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponent.scala]: I'd reduce these two `val`s into a single call, by adding the `.sum` right into the first:
```
val statusCounts:Map[String, Int] = states.getOrElse(submissionRec.id, Seq.empty).map(x => x.workflowStatus -> x.count).groupBy(_._1) map {
  case (status, statusToCount) => status -> statusToCount.map(_._2).sum
}
```
and, if you feel like diving into a fairly minimal use of cats, you can replace the groupBy/map/case/sum:
```
import cats.implicits._

val statusCounts:Map[String, Int] = states.getOrElse(submissionRec.id, Seq.empty).map(x => x.workflowStatus -> x.count)
    .foldLeft(Map.empty[String,Int])(_|+|_)
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -1393,14 +1393,26 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
   }
 
-  def listSubmissions(workspaceName: WorkspaceName): Future[PerRequestMessage] =
-    dataSource.inTransaction { dataAccess =>
+  def listSubmissions(workspaceName: WorkspaceName): Future[PerRequestMessage] = {
+    val costlessSubmissionListFuture = dataSource.inTransaction { dataAccess =>
       withWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevels.Read, dataAccess) { workspaceContext =>
-        dataAccess.submissionQuery.listWithSubmitter(workspaceContext)
-          .map(RequestComplete(StatusCodes.OK, _))
+        dataAccess.submissionQuery.listWithSubmitter(workspaceContext).flatMap(DBIO.successful)
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: I think the `.flatMap(DBIO.successful)` is extraneous here - just remove it

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -1393,14 +1393,26 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+        toFutureTry(submissionCostService.getWorkflowCosts(costlessSubmission.workflowIds, workspaceName.namespace)) map {
+          case Failure(ex) =>
+            logger.error(s"Unable to get cost data for submission ${costlessSubmission.submissionId}", ex)
+            costlessSubmission
+          case Success(costs) =>
+            costlessSubmission.copy(cost = Some(costs.values.sum))
+        }
+      }) map (RequestComplete(StatusCodes.OK, _))
+    }
+  }
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: I don't like this approach. This is making N request to BQ, where N is the number of submissions in the workspace. Could we instead make a single BQ request for all workflowIds, then parse/extract/annotate the costs in Scala? I worry about performance for N queries.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/SubmissionCostServiceSpec.scala [hunk 4]
```diff
@@ -23,9 +23,15 @@ class SubmissionCostServiceSpec extends FlatSpec with RawlsTestUtils {
...
       Await.result(submissionCostService.extractWorkflowCostResults(rows), 1 minute)
     }
   }
 
+  it should "bypass BigQuery with no workflow IDs" in {
+    val expected = Map.empty
+    assertResult(expected) {
+      Await.result(submissionCostService.getWorkflowCosts(Seq.empty, "test"), 1 minute)
+    }
+  }
```
Comment (davidangb) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/SubmissionCostServiceSpec.scala]: this doesn't actually test that we're bypassing BQ - it just tests that we get an empty map in the response. We coul easily have made a request to BQ but returned nothing.

**Review by davidangb — COMMENTED** [review]
comments inline!

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/SubmissionCostServiceSpec.scala (hunk 4 above)
Comment (ghost) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/SubmissionCostServiceSpec.scala]: I'd actually like to argue that it does test bypassing BQ. My reason being that `MockGoogleBigQueryDAO`, which is what this `submissionCostService` will hit, throws an exception if the parameters passed to `startParameterizedQuery` are not equal to the fields `testProject`, `testParamQuery`, `testParameters` and `testParameterMode` in `MockGoogleBigQueryDAO`. Since this is version of `SubmissionCostService` isn't passing those exact parameters, we'll always see an exception if we try to hit BigQuery in this spec, unless we override `SubmissionCostService#executeWorkflowCostQuery`. I'm interested to hear what you think about this.

Comment (jbuabbud) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/SubmissionCostServiceSpec.scala]: I rescind my previous comment.

Comment (davidangb) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/SubmissionCostServiceSpec.scala]: you're absolutely right! I didn't follow this all the way down. Could you please add comments explaining this, since I expect it wouldn't be obvious to other readers too?

Comment (ghost) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/SubmissionCostServiceSpec.scala]: Done!

**Review by jmthibault79 — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 5]
```diff
@@ -1393,14 +1393,31 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+    }
+
+    toFutureTry(costMapFuture) flatMap {
+      case Failure(ex) =>
+        logger.error("Unable to get cost data from BigQuery", ex)
+        Future.successful(RequestComplete(StatusCodes.OK, costlessSubmissionsFuture))
+      case Success(costMap) =>
+        costlessSubmissionsFuture map { costlessSubmissions =>
+          val costedSubmissions = costlessSubmissions map { costlessSubmission =>
+            costlessSubmission.copy(cost = Some(costlessSubmission.workflowIds.flatMap(id => costMap.get(id).value).sum))
```
Comment (davidangb) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: what does `value` do here?  You should be able to do just `costlessSubmission.workflowIds.flatMap(costMap.get).sum` (note I've also removed the `id =>` and the shorthand `_` since `get` is a single-argument function.

If you wanted to keep using `map`, you could do `costlessSubmission.workflowIds.map(costMap.getOrElse(_,0).sum`

**Review by davidangb — COMMENTED** [review]
one fairly minor comment inline. Have you run this and manually looked at the results?

**Comment by ghost** [comment]
@davidangb screen shot 2018-05-24 at 2 11 26 pm

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 5 above)
Comment (ghost) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: 🤦‍♂️ This was my hacky way of fixing a bug that was really caused by having `Some().sum`. Going with the first suggestion you made. Thanks for the comment 😄

**Review by davidangb — APPROVED** [review]
:+1: thanks!
